File: src/test/baekjoon/python/binarysearch/KthNumber.py
N = int(input())
k = int(input())
start, end = 1, k
answer = 0
while start <= end:
    count = 0
    mid = (start + end) // 2
    for i in range(1, N + 1):
        count += min(mid // i, N)
    if count >= k:
        answer = mid
        end = mid - 1
    else:
        start = mid + 1
print(answer)

File: src/test/baekjoon/java/binarysearch/KthNumber.java
package test.baekjoon.java.binarysearch;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class KthNumber {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());
        int k = Integer.parseInt(br.readLine());

        int start = 1;
        int end = k;
        int answer = 0;

        while(start <= end) {
            int count = 0;
            int mid = (start + end) / 2;

            for(int i = 1; i <= N; i++) {
                count += Math.min((mid / i), N);
            }

            if(count >= k) {
                answer = mid;
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }

        System.out.println(answer);
    }
}
